src/wallet/mod.rs

```diff
@@ -387,60 +394,31 @@ where
         Ok(new_addresses_cached)
     }
 
-    /// Return vector of n unused addresses from the [`KeychainKind`].
-    /// If less than n unused addresses are returned, the rest will be populated by new addresses.
-    /// The unused addresses returned are in order of oldest in keychain first, with increasing index.
-    pub fn get_batch_unused_addresses(
-        &self,
-        n: usize,
-        from_front: bool,
-        keychain: KeychainKind,
-    ) -> Result<Vec<AddressInfo>, Error> {
+    /// Return set of unused script indexes for the [`KeychainKind`].
+    pub fn get_unused_script_indexes(&self, keychain: KeychainKind) -> Result<Vec<u32>, Error> {
         let script_pubkeys = self
             .database
             .borrow()
             .iter_script_pubkeys(Some(keychain))
             .unwrap_or_else(|_| vec![]);
-
         let current_address_index = self.fetch_index(keychain)? as usize;
-        let check_indexes = if from_front {
-            (0..=current_address_index).collect::<Vec<_>>()
-        } else {
-            (0..=current_address_index).rev().collect::<Vec<_>>()
-        };
+        let txns = self.list_transactions(true).unwrap_or_else(|_| vec![]);
 
-        let mut unused_addresses = vec![];
-        for i in check_indexes {
-            // if we have made a pubkey at this index, check whether the address has been used.
-            if i < script_pubkeys.len() {
-                let script_pk = &script_pubkeys[i];
-                if self.has_address_been_used(script_pk) {
-                    continue;
-                }
-            }
-            if let Ok(unused_address) = self
-                .get_descriptor_for_keychain(keychain)
-                .as_derived(i as u32, &self.secp)
-                .address(self.network)
-                .map(|address| AddressInfo {
-                    address,
-                    index: i as u32,
-                    keychain,
-                })
-                .map_err(|_| Error::ScriptDoesntHaveAddressForm)
-            {
-                unused_addresses.push(unused_address);
-            }
+        let txn_scripts: HashSet<Script> = txns
+            .iter()
+            .flat_map(|tx_details| tx_details.transaction.as_ref())
+            .flat_map(|tx| tx.output.iter())
+            .map(|o| o.script_pubkey.clone())
+            .collect();
 
-            if unused_addresses.len() >= n {
-                break;
+        let mut scripts_not_used = vec![];
+        for i in 0..=current_address_index {
+            if (i < script_pubkeys.len()) && txn_scripts.contains(&script_pubkeys[i]) {
+            } else {
+                scripts_not_used.push(i as u32);
```

Comment:
this `if` is kind of awkward.Can't we just iterate over `script_pubkeys.enumerate()` and push `i` if not `contains`.

Reply:
So one reason i iterated over `0..=current_address_index` was because the script pubkeys can be empty for a new database, but address index 0. Then `script_pubkeys.enumerate()` misses index 0 and returns nothing. I think we want to return `[0]` on a fresh wallet

I can add this manually after the enumerate but it's not so nice
```
        if script_pubkeys.len() == 0 {
            scripts_not_used.push(0);
        }
```

Related to earlier comment, the other reason is it seems possible for more scripts in the database to be populated further than the current address index.
![image](https://user-images.githubusercontent.com/24557779/180732519-f0265519-f13c-4ea5-801b-e9e0ea7f0d72.png)

Can handle this upper limit with `take(current_address_index + 1)`.

Also worried about having less scripts than the `current_address_index`, gaps possible?

Reply:
The `take(current_address_index + 1)` is OK because it never gives an exception even if the index is larger than the length of `script_pubkeys`.

Reply:
Ok now I get it. Yeah I would do the `take` thing. Are you sure that `current_address_index` is not the one *after* the one you are looking for i.e. you should be iterating `.take(current_address_index)` and not `current_address_index + 1`. i.e.  the "current" address is really the "next address to use".

Reply:
So `current_address_index` is the index of the address last derived. When the `current_address_index` is `0` we need to take `1` address and check if it has been used

Reply:
yikes that sucks. :vomiting_face: